src/gt4py/next/otf/compiled_program.py

```diff
@@ -0,0 +1,287 @@
+# GT4Py - GridTools Framework
+#
+# Copyright (c) 2014-2024, ETH Zurich
+# All rights reserved.
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+
+from __future__ import annotations
+
+import concurrent.futures
+import dataclasses
+import functools
+import itertools
+from collections.abc import Sequence
+from typing import Any, TypeAlias
+
+from gt4py._core import definitions as core_defs
+from gt4py.eve import extended_typing, utils as eve_utils
+from gt4py.next import backend as gtx_backend, common, config, errors
+from gt4py.next.ffront import stages as ffront_stages, type_specifications as ts_ffront
+from gt4py.next.otf import arguments, stages
+from gt4py.next.type_system import type_info, type_specifications as ts, type_translation as tt
+
+
+# TODO(havogt): We would like this to be a ProcessPoolExecutor, which requires (to decide what) to pickle.
+_async_compilation_pool = concurrent.futures.ThreadPoolExecutor(max_workers=config.BUILD_JOBS)
+
+ScalarOrTupleOfScalars: TypeAlias = extended_typing.TOrNestedTuple[core_defs.Scalar]
+
+
+# TODO(havogt): We should look at the hash related codepaths again when tuning Python overheads.
+@dataclasses.dataclass
+class _CompiledProgramsKey:
+    values: tuple[ScalarOrTupleOfScalars, ...]  # in order of static_params
+    offset_provider_type: common.OffsetProviderType
+
+    def __hash__(self) -> int:
+        assert common.is_offset_provider_type(self.offset_provider_type)
+        return hash((self.values, frozenset(self.offset_provider_type.items())))
+
+
+def _convert_to_type(
+    value: ScalarOrTupleOfScalars, type_: ts.TupleType | ts.ScalarType
+) -> ScalarOrTupleOfScalars:
+    if isinstance(type_, ts.ScalarType):
+        return tt.as_dtype(type_).scalar_type(value)
+    else:
+        assert isinstance(type_, ts.TupleType)
+        assert isinstance(value, tuple)
+        assert all(isinstance(t, (ts.ScalarType, ts.TupleType)) for t in type_.types)
+        return tuple(_convert_to_type(v, t) for v, t in zip(value, type_.types))  # type: ignore[arg-type] # checked in assert
+
+
+def _validate_types(
+    program_name: str,
+    static_args: dict[str, ScalarOrTupleOfScalars],
+    program_type: ts_ffront.ProgramType,
+) -> None:
+    param_types = program_type.definition.pos_or_kw_args
+
+    type_errors = []
+    for name, type_ in param_types.items():
+        if name in static_args:
+            if not type_info.is_type_or_tuple_of_type(type_, ts.ScalarType):
+                type_errors.append(f"'{name}' with type '{type_}' cannot be static.")
```

Comment:
Unless you prefer the explicit loop version for debugging, I'd rather use a list-comprehension:
```suggestion
    type_errors = [
        f"'{name}' with type '{type_}' cannot be static."
        for name, type_ in param_types.items()
        if name in static_args and not type_info.is_type_or_tuple_of_type(type_, ts.ScalarType)
    ]
```
Or, actually, isn't that equivalent to just iterating `static_args` directly?

```suggestion
    type_errors = [
        f"'{name}' with type '{type_}' cannot be static."
        for name in static_arts
        if not type_info.is_type_or_tuple_of_type(param_types[name], ts.ScalarType)
    ]
```